src/backend/joanie/tests/core/test_api_admin_orders.py

```diff
@@ -1525,3 +1530,323 @@ def test_api_admin_orders_generate_certificate_when_order_is_not_ready_for_gradi
+        """
+        admin = factories.UserFactory(is_staff=True, is_superuser=True)
+        self.client.login(username=admin.username, password="password")
+        order = factories.OrderGeneratorFactory(state=enums.ORDER_STATE_PENDING)
+        order.flow.cancel()
+
+        response = self.client.post(f"/api/v1.0/admin/orders/{order.id}/refund/")
+
+        self.assertEqual(response.status_code, HTTPStatus.BAD_REQUEST)
+
+    def test_api_admin_orders_refund_an_order(self):
+        """
+        Authenticated admin users should be able to refund an order when the state is `canceled`.
+        Once the treatment is done, the order's state should be `refund` and the installment
+        state should be set to `refunded`.
+        """
+        backend = DummyPaymentBackend()
+        request_factory = APIRequestFactory()
+        admin = factories.UserFactory(is_staff=True, is_superuser=True)
+        self.client.login(username=admin.username, password="password")
+        order = factories.OrderGeneratorFactory(
+            state=enums.ORDER_STATE_PENDING, product__price=100
+        )
+        # Create the payment
+        payment_id = backend.create_payment(
+            order=order,
+            installment=order.payment_schedule[0],
+            billing_address=order.main_invoice.recipient_address.to_dict(),
+        )["payment_id"]
+
+        # Notify that payment has been paid
+        request = request_factory.post(
+            reverse("payment_webhook"),
+            data={
+                "id": payment_id,
+                "type": "payment",
+                "state": "success",
+                "installment_id": order.payment_schedule[0]["id"],
+            },
+            format="json",
+        )
+        request.data = json.loads(request.body.decode("utf-8"))
+        backend.handle_notification(request)
+
+        # The installment's state should be in state 'paid
+        order.refresh_from_db()
+        self.assertEqual(order.payment_schedule[0]["state"], enums.PAYMENT_STATE_PAID)
+
+        order.flow.cancel()
+
+        response = self.client.post(f"/api/v1.0/admin/orders/{order.id}/refund/")
+
+        order.refresh_from_db()
+        self.assertEqual(response.status_code, HTTPStatus.ACCEPTED)
+        self.assertEqual(order.state, enums.ORDER_STATE_REFUND)
+
+        # Notify that payment has been refunded
+        request = request_factory.post(
+            reverse("payment_webhook"),
+            data={
+                "id": payment_id,
+                "type": "refund",
+                "amount": order.payment_schedule[0]["amount"].sub_units,
+                "installment_id": order.payment_schedule[0]["id"],
+            },
+            format="json",
+        )
+        request.data = json.loads(request.body.decode("utf-8"))
+
+        backend.handle_notification(request)
+
+        order.refresh_from_db()
+        self.assertEqual(
```

Comment:
Also all installments should have a `refunded` state ?

Reply:
Only the paid installment should be set to `refunded`.